pkg/chains/formats/sbom/sbom.go

```diff
@@ -0,0 +1,106 @@
+/*
+Copyright 2023 The Tekton Authors
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sbom
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	slsa "github.com/in-toto/in-toto-golang/in_toto/slsa_provenance/v0.2"
+	"github.com/tektoncd/chains/pkg/chains/objects"
+	"k8s.io/client-go/kubernetes"
+)
+
+const maxSBOMBytes = 5 * 1024 * 1024 // 5 Megabytes
+
+func GenerateAttestation(ctx context.Context, kc kubernetes.Interface, builderID string, sbom *objects.SBOMObject) (interface{}, error) {
+	subject := []intoto.Subject{
+		{Name: sbom.GetImageURL(), Digest: toDigestSet(sbom.GetImageDigest())},
+	}
+
+	data, err := getData(ctx, kc, sbom)
+	if err != nil {
+		return nil, err
+	}
+
+	att := intoto.Statement{
+		StatementHeader: intoto.StatementHeader{
+			Type:          intoto.StatementInTotoV01,
+			PredicateType: sbom.GetSBOMFormat(),
+			Subject:       subject,
+		},
+		Predicate: data,
+	}
+	return att, nil
+}
+
+func getData(ctx context.Context, kc kubernetes.Interface, sbom *objects.SBOMObject) (json.RawMessage, error) {
+	opt, err := sbom.OCIRemoteOption(ctx, kc)
+	if err != nil {
+		return nil, err
+	}
+
+	uri := sbom.GetSBOMURL()
+	ref, err := name.ParseReference(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	image, err := remote.Image(ref, opt)
+	if err != nil {
+		return nil, err
+	}
+
+	layers, err := image.Layers()
+	if err != nil {
+		return nil, err
+	}
+
+	if len(layers) != 1 {
+		return nil, fmt.Errorf("expected exactly 1 layer in sbom image %s, found %d", uri, len(layers))
+	}
+
+	layer, err := layers[0].Uncompressed()
+	if err != nil {
+		return nil, err
+	}
+	defer layer.Close()
+
+	var blob bytes.Buffer
+	if _, err := io.Copy(&blob, io.LimitReader(layer, maxSBOMBytes)); err != nil {
+		return nil, err
+	}
+
+	var data json.RawMessage
+	if err := json.Unmarshal(blob.Bytes(), &data); err != nil {
+		return nil, err
+	}
```

Comment:
This would also solve the 5MB image problem, since we'd sign the the manifest rather than the direct SBOM content.

Reply:
@wlynch, I don' think we should be signing the .sbom images, mostly because of the problem outlined in [sigstore/cosign#2755](https://github.com/sigstore/cosign/issues/2755). If someone does want to sign those sbom images, then they can simply use the existing `IMAGE_URL` result type hint. Chains will happily sign those.

Using attestations to store SBOMs seems like the right approach here, and embedding the SBOM in the attestation payload seems correct as well.

Good call on documenting that this only supports JSON format. I think with a little tweak, it could support anything though.

Reply:
I think this is addressed now.